File: java/3171. Find Subarray With Bitwise AND Closest to K.java
class Solution {
    public int minimumDifference(int[] nums, int k) {
        int[] bitFreq = new int[32];
        int left = 0;
        int answer = Integer.MAX_VALUE;
        for (int i : nums) answer = Math.min(answer, Math.abs(i-k));
        for (int right = 0; right < nums.length; right++) {
            add(bitFreq, nums[right]);
            int num = getNumber(bitFreq, right-left+1);
            answer = Math.min(answer, Math.abs(num-k));
            while (left <= right && num < k) {
                remove(bitFreq, nums[left]);
                left++;
                num = getNumber(bitFreq, right-left+1);
                answer = Math.min(answer, Math.abs(num - k));
            }
        }
        return answer;
    }
    private int getNumber(int[] freq, int n) {
        int answer = 0;
        for (int i = 0; i < 32; i++) {
            if (freq[i] == n) {
                answer ^= (1 << i);
            }
        }
        return answer;
    }
    private void add(int[] freq, int n) {
        for (int i = 0; i < 32; i++) {
            if ((n & (1 << i)) > 0) {
                freq[i]++;
            }
        }
    }
    private void remove(int[] freq, int n) {
        for (int i = 0; i < 32; i++) {
            if ((n & (1 << i)) > 0) {
                freq[i]--;
            }
        }
    }
}

File: go/3171. Find Subarray With Bitwise AND Closest to K.go
func minimumDifference(nums []int, k int) int {
    freq := make([]int, 32)
    answer := math.MaxInt32
    left := 0
    for right := 0; right < len(nums); right++ {
        add(freq, nums[right])
        curr := get(freq, right-left+1)
        answer = min(answer, abs(curr-k))
        for left <= right && curr < k {
            remove(freq, nums[left])
            left += 1
            curr = get(freq, right-left+1)
            answer = min(answer, abs(curr-k))
        }
    }
    return answer
}

func abs(i int) int {
    if i < 0 {
        return -i
    }
    return i
}

func add(freq []int, n int) {
    for i := 0; i < 32; i++ {
        if (n & (1 << i)) > 0 {
            freq[i] += 1
        }
    }
}

func remove(freq []int, n int) {
    for i := 0; i < 32; i++ {
        if (n & (1 << i)) > 0 {
            freq[i] -= 1
        }
    }
}

func get(freq []int, n int) int {
    answer := 0
    for i := 0; i < 32; i++ {
        if freq[i] == n {
            answer ^= (1 << i)
        }
    }
    return answer
}
